cpp/src/reshape/byte_cast.cu

```diff
@@ -31,123 +33,160 @@
 #include <thrust/iterator/constant_iterator.h>
 #include <thrust/iterator/counting_iterator.h>
 
+#include <type_traits>
+
 namespace cudf {
 namespace detail {
 namespace {
-struct byte_list_conversion {
-  /**
-   * @brief Function object for converting primitive types and string columns to lists of bytes.
-   */
-  template <typename T>
-  std::enable_if_t<!std::is_integral_v<T> and !is_floating_point<T>(), std::unique_ptr<column>>
-  operator()(column_view const&,
-             flip_endianness,
-             rmm::cuda_stream_view,
-             rmm::mr::device_memory_resource*) const
+
+// Data type of the output data column after conversion.
+data_type constexpr output_type{data_type{type_id::UINT8}};
```

Comment:
We don't really use "east constexpr" anywhere in our codebase. Many of the reasons that east const makes sense don't really make sense for constexpr, particularly when it comes to the confusion around pointers. For instance `data_type constexpr* output_type` does not in fact mean "pointer to constant data_type", it means "constant pointer to data_type". For instance, check out this example: https://godbolt.org/z/o4YaEhrsE.